src/linear_samplers/block_col_sample_wo_replacement.jl

```diff
@@ -0,0 +1,107 @@
+- `population::Union{Vector{Int64}, Nothing}`, buffer array to hold the vector `collect(1:size(A)[2])` in the `sample` function.
+- `col_sampled::Union{Vector{Int64}, Nothing}`, buffer array to hold index of columns sampled from `A`.
+- `S::Union{Matrix{Int64}, Nothing}`, buffer array to hold sketching matrix `S`.
+
+Calling `LinSysBlkColSelectWoReplacement()` defaults to `LinSysBlkColSelectWoReplacement(2, nothing, nothing, nothing, nothing)`.
+
+An additional constructor is provided with the keyword arguments `block_size` and `probability`.
+"""
+mutable struct LinSysBlkColSelectWoReplacement <: LinSysVecColSelect
+    block_size::Int64
+    probability::Union{Weights, Vector{Float64}, Nothing}
+    population::Union{Vector{Int64}, Nothing}
+    col_sampled::Union{Vector{Int64}, Nothing}
+    S::Union{Matrix{Int64}, Nothing}
+end
+
+function LinSysBlkColSelectWoReplacement(;block_size = 2, probability = nothing)
+    return LinSysBlkColSelectWoReplacement(block_size, probability, nothing, nothing, nothing)
+end
+
+# Common sample interface for linear systems
+function sample(
+    type::LinSysBlkColSelectWoReplacement,
+    A::AbstractArray,
+    b::AbstractVector,
+    x::AbstractVector,
+    iter::Int64
+)
+
+    # Error checking and initialization on first iteration
+    if iter == 1
+        ncol = size(A)[2]
+        if type.block_size <= 0
+            throw(DomainError("block_size is 0 or negative!")) 
+        end
+        
+        if type.block_size > ncol
+            throw(DomainError("block_size is larger than the number of columns in A! Cannot select this amount without replacement!"))
+        end
+
+        # initialization buffer arrays
+        type.S = Matrix{Int64}(undef, ncol, type.block_size)
+        type.col_sampled = Vector{Int64}(undef, type.block_size)
+
+        # error checking on probability vector, possible initialization and type conversion
+        if isnothing(type.probability)
+            type.probability = Weights(repeat([1/ncol], outer = ncol))
+        elseif isa(type.probability, Vector) || isa(type.probability, Weights)
+            # check that probability is a valid distribution on the rows
+            if !(sum(type.probability) ≈ 1)
+                throw(DomainError("Elements in probability do not sum to one!"))
+            elseif sum(type.probability .>= 0) != size(type.probability)[1]
+                throw(DomainError("Not all probabilities are non-negative in probability!"))
+            elseif size(type.probability)[1] != size(A)[2]
+                throw(DimensionMismatch("Length of probability vector is smaller than the number of columns in A!"))
+            elseif sum(type.probability .> 0) < type.block_size
+                throw(DimensionMismatch("Not enough non-zero probabilities in probability to select the required number of columns! Either decrease block_size, or reweight probability."))
```

Comment:
Really like these warnings maybe place `probability` in quotation marks to clarify that you are referring to the variable.